File: xtquant/agent/cc_switch.py
"""CC Switch — Anthropic ↔ OpenAI API proxy for Claude Code / Cursor / Codex.

Runs as a local HTTP proxy, translating Anthropic Messages API requests
to OpenAI Chat Completions format so you can use DeepSeek / Qwen / etc.
"""

import json
import uuid
import logging
import threading
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler

logger = logging.getLogger("xtquant.agent.cc_switch")

# ── Global state ──
_lock = threading.RLock()
_config: dict = {
    "port": 12435,
    "target_url": "https://api.deepseek.com/v1/chat/completions",
    "target_model": "deepseek-chat",
    "api_key": "",
    "running": False,
}
_server: HTTPServer | None = None
_server_thread: threading.Thread | None = None


def get_status() -> dict:
    with _lock:
        return {
            "running": _config.get("running", False),
            "port": _config.get("port", 12435),
            "target_url": _config.get("target_url", ""),
            "target_model": _config.get("target_model", ""),
            "has_api_key": bool(_config.get("api_key")),
        }


def configure(port: int = None, target_url: str = None,
              target_model: str = None, api_key: str = None) -> dict:
    with _lock:
        if port is not None:
            _config["port"] = port
        if target_url is not None:
            _config["target_url"] = target_url
        if target_model is not None:
            _config["target_model"] = target_model
        if api_key is not None:
            _config["api_key"] = api_key
    return get_status()


def _translate_anthropic_to_openai(body: dict, target_model: str = "deepseek-chat") -> dict:
    """Convert Anthropic Messages request to OpenAI Chat Completions."""
    messages = []

    # System prompt → system message
    system = body.get("system", "")
    if system:
        if isinstance(system, list):
            system = "\n".join(s.get("text", "") for s in system if isinstance(s, dict))
        messages.append({"role": "system", "content": system})

    # Messages
    for msg in body.get("messages", []):
        role = msg.get("role", "")
        content = msg.get("content", "")
        if isinstance(content, list):
            parts = []
            for block in content:
                if isinstance(block, dict):
                    if block.get("type") == "text":
                        parts.append(block.get("text", ""))
                    elif block.get("type") == "image":
                        parts.append("[image]")
                    elif block.get("type") == "tool_use":
                        parts.append(json.dumps(block, ensure_ascii=False))
                    elif block.get("type") == "tool_result":
                        parts.append(json.dumps(block, ensure_ascii=False))
            content = "\n".join(parts)
        messages.append({"role": role, "content": content})

    openai_body = {
        "model": target_model,
        "messages": messages,
    }

    max_tokens = body.get("max_tokens", 4096)
    openai_body["max_tokens"] = min(max_tokens, 4096)

    if body.get("temperature") is not None:
        openai_body["temperature"] = body["temperature"]

    tools = body.get("tools")
    if tools:
        openai_body["tools"] = tools
        tool_choice = body.get("tool_choice")
        if tool_choice:
            openai_body["tool_choice"] = tool_choice

    # Strictly force tool mode back to the proxy
    openai_body["tool_choice"] = "auto"

    return openai_body


def _translate_openai_to_anthropic(data: dict, orig_model: str) -> dict:
    """Convert OpenAI Chat Completion response to Anthropic Messages."""
    choice = data.get("choices", [{}])[0]
    msg = choice.get("message", {})
    content_text = msg.get("content") or ""
    finish = choice.get("finish_reason", "stop")

    # Build Anthropic content blocks
    content_blocks = []
    if content_text:
        content_blocks.append({"type": "text", "text": content_text})

    # Tool calls
    tool_calls = msg.get("tool_calls", []) or []
    for tc in tool_calls:
        func = tc.get("function", {})
        try:
            args = json.loads(func.get("arguments", "{}"))
        except (json.JSONDecodeError, TypeError):
            args = {}
        content_blocks.append({
            "type": "tool_use",
            "id": tc.get("id", "toolu_" + uuid.uuid4().hex[:8]),
            "name": func.get("name", ""),
            "input": args,
        })

    stop_map = {"stop": "end_turn", "tool_calls": "tool_use", "length": "max_tokens"}
    stop_reason = stop_map.get(finish, "end_turn")

    usage = data.get("usage", {})

    return {
        "id": "msg_" + uuid.uuid4().hex[:12],
        "type": "message",
        "role": "assistant",
        "content": content_blocks,
        "model": orig_model,
        "stop_reason": stop_reason,
        "stop_sequence": None,
        "usage": {
            "input_tokens": usage.get("prompt_tokens", 0),
            "output_tokens": usage.get("completion_tokens", 0),
        },
    }


class _ProxyHandler(BaseHTTPRequestHandler):

    def log_message(self, fmt, *args):
        logger.debug("CC_SWITCH: " + fmt, *args)

    def _send(self, code: int, body: dict | str):
        data = json.dumps(body, ensure_ascii=False) if isinstance(body, dict) else body
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(data.encode("utf-8"))

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "*")
        self.end_headers()

    def do_GET(self):
        if self.path.startswith("/v1/models"):
            # Forward models list to target
            self._proxy_to_target("GET")
        else:
            self._send(404, {"error": "not found"})

    def do_POST(self):
        if self.path.startswith("/v1/messages"):
            self._handle_messages()
        else:
            self._send(404, {"error": "not found"})

    def _handle_messages(self):
        length = int(self.headers.get("Content-Length", 0))
        raw = self.rfile.read(length)
        try:
            body = json.loads(raw)
        except json.JSONDecodeError:
            self._send(400, {"error": "invalid json"})
            return

        orig_model = body.get("model", "unknown")
        with _lock:
            api_key = _config.get("api_key", "")
            target_url = _config.get("target_url", "")
            target_model = _config.get("target_model", "deepseek-chat")

        if not api_key:
            self._send(500, {"error": "CC Switch API key not configured"})
            return

        openai_body = _translate_anthropic_to_openai(body, target_model)
        req = urllib.request.Request(
            target_url,
            data=json.dumps(openai_body).encode("utf-8"),
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
        )

        try:
            resp = urllib.request.urlopen(req, timeout=120)
            resp_data = json.loads(resp.read().decode("utf-8"))
            anthropic_resp = _translate_openai_to_anthropic(resp_data, orig_model)
            self._send(200, anthropic_resp)
        except urllib.error.HTTPError as e:
            err_body = e.read().decode("utf-8", errors="replace")
            logger.error("CC_SWITCH upstream error %s: %s", e.code, err_body[:300])
            self._send(e.code, {"error": f"upstream: {e.code} {err_body[:200]}"})
        except Exception as e:
            logger.error("CC_SWITCH proxy error: %s", e)
            self._send(502, {"error": str(e)[:300]})

    def _proxy_to_target(self, method: str):
        """Simple GET forward for /v1/models etc."""
        with _lock:
            api_key = _config.get("api_key", "")
            target = _config.get("target_url", "").replace("/chat/completions", "/models")
        req = urllib.request.Request(
            target,
            headers={"Authorization": f"Bearer {api_key}"} if api_key else {},
        )
        try:
            resp = urllib.request.urlopen(req, timeout=30)
            data = resp.read()
            self.send_response(resp.status)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(data)
        except Exception as e:
            self._send(502, {"error": str(e)[:200]})


def start() -> dict:
    global _server, _server_thread
    with _lock:
        if _config.get("running"):
            return {"status": "already_running", **get_status()}

        port = _config.get("port", 12435)
        try:
            _server = HTTPServer(("127.0.0.1", port), _ProxyHandler)
            _server_thread = threading.Thread(target=_server.serve_forever, daemon=True)
            _server_thread.start()
            _config["running"] = True
            logger.info("CC_SWITCH started on http://127.0.0.1:%d", port)
            return {"status": "started", **get_status()}
        except OSError as e:
            return {"status": "error", "error": str(e), **get_status()}


def stop() -> dict:
    global _server, _server_thread
    with _lock:
        if _server:
            try:
                _server.shutdown()
            except Exception:
                pass
            _server = None
        _server_thread = None
        _config["running"] = False
        logger.info("CC_SWITCH stopped")
        return {"status": "stopped", **get_status()}

File: gateway/internal/agent/ccswitch.go
package agent

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
)

// ── CC Switch: Anthropic ↔ OpenAI Format Converter ──

// CCSwitch is an HTTP proxy that converts between Anthropic Messages API
// and OpenAI Chat Completions API formats. This allows tools like Claude Code
// or Cursor to connect to any OpenAI-compatible backend (DeepSeek, Qwen, etc.).
type CCSwitch struct {
	TargetURL  string
	TargetKey  string
	TargetModel string
}

// NewCCSwitch creates a new format converter proxy.
func NewCCSwitch(targetURL, apiKey, model string) *CCSwitch {
	return &CCSwitch{
		TargetURL:  strings.TrimRight(targetURL, "/"),
		TargetKey:  apiKey,
		TargetModel: model,
	}
}

// HandleAnthropicRequest processes an Anthropic-formatted request and converts
// it to an OpenAI-compatible request, then converts the response back.
func (c *CCSwitch) HandleAnthropicRequest(w http.ResponseWriter, r *http.Request) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, err.Error(), 400)
		return
	}
	defer r.Body.Close()

	var anthropicReq map[string]any
	if err := json.Unmarshal(body, &anthropicReq); err != nil {
		http.Error(w, "invalid JSON", 400)
		return
	}

	messages, _ := anthropicReq["messages"].([]any)
	systemMsg, _ := anthropicReq["system"].(string)
	model, _ := anthropicReq["model"].(string)
	maxTokens, _ := anthropicReq["max_tokens"].(float64)
	stream, _ := anthropicReq["stream"].(bool)

	if model == "" {
		model = c.TargetModel
	}

	// Convert to OpenAI format
	openAIReq := map[string]any{
		"model":    model,
		"messages": convertAnthropicToOpenAIMessages(messages, systemMsg),
	}
	if maxTokens > 0 {
		openAIReq["max_tokens"] = int(maxTokens)
	}
	// Anthropic defaults to streaming; OpenAI uses stream per request
	if stream {
		openAIReq["stream"] = true
	}

	reqBody, _ := json.Marshal(openAIReq)
	req, err := http.NewRequest("POST", c.TargetURL+"/v1/chat/completions", bytes.NewReader(reqBody))
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.TargetKey)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		http.Error(w, err.Error(), 502)
		return
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)

	if stream {
		// Convert streaming SSE response
		c.writeStreamResponse(w, respBody)
		return
	}

	// Convert non-streaming response back to Anthropic format
	var openAIResp map[string]any
	json.Unmarshal(respBody, &openAIResp)

	anthropicResp := convertOpenAIToAnthropicResponse(openAIResp, model)

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(anthropicResp)
}

// HandleOpenAIRequest passes through or converts an OpenAI request.
func (c *CCSwitch) HandleOpenAIRequest(w http.ResponseWriter, r *http.Request) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, err.Error(), 400)
		return
	}
	defer r.Body.Close()

	var req map[string]any
	json.Unmarshal(body, &req)

	// Override model if target is set
	if c.TargetModel != "" {
		req["model"] = c.TargetModel
	}

	reqBody, _ := json.Marshal(req)
	proxyReq, err := http.NewRequest("POST", c.TargetURL+"/v1/chat/completions", bytes.NewReader(reqBody))
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	proxyReq.Header.Set("Content-Type", "application/json")
	proxyReq.Header.Set("Authorization", "Bearer "+c.TargetKey)

	resp, err := http.DefaultClient.Do(proxyReq)
	if err != nil {
		http.Error(w, err.Error(), 502)
		return
	}
	defer resp.Body.Close()

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(resp.StatusCode)
	io.Copy(w, resp.Body)
}

// HandleModels returns model list in OpenAI format.
func (c *CCSwitch) HandleModels(w http.ResponseWriter, r *http.Request) {
	resp := map[string]any{
		"object": "list",
		"data": []map[string]any{
			{
				"id":       c.TargetModel,
				"object":   "model",
				"owned_by": "ccswitch",
			},
		},
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resp)
}

func (c *CCSwitch) writeStreamResponse(w http.ResponseWriter, openAIBody []byte) {
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")

	// Parse SSE from OpenAI and convert each chunk to Anthropic SSE format
	lines := strings.Split(string(openAIBody), "\n")
	for _, line := range lines {
		if !strings.HasPrefix(line, "data: ") {
			continue
		}
		data := strings.TrimPrefix(line, "data: ")
		if data == "[DONE]" {
			fmt.Fprintf(w, "event: message_stop\ndata: {\"type\":\"message_stop\"}\n\n")
			break
		}

		var chunk map[string]any
		if json.Unmarshal([]byte(data), &chunk) != nil {
			continue
		}

		anthropicChunk := convertOpenAIChunkToAnthropic(chunk)
		chunkJSON, _ := json.Marshal(anthropicChunk)
		fmt.Fprintf(w, "event: content_block_delta\ndata: %s\n\n", string(chunkJSON))
	}
	w.(http.Flusher).Flush()
}

// ── Message Conversion ──

func convertAnthropicToOpenAIMessages(messages []any, systemMsg string) []map[string]any {
	var openAIMsgs []map[string]any
	if systemMsg != "" {
		openAIMsgs = append(openAIMsgs, map[string]any{
			"role":    "system",
			"content": systemMsg,
		})
	}
	for _, msg := range messages {
		m, _ := msg.(map[string]any)
		role, _ := m["role"].(string)
		if role == "assistant" {
			role = "assistant"
		} else {
			role = "user"
		}
		content := m["content"]
		// Anthropic content can be array or string
		if contentArr, ok := content.([]any); ok {
			var textParts []string
			for _, block := range contentArr {
				if blockMap, ok := block.(map[string]any); ok {
					if text, ok := blockMap["text"].(string); ok {
						textParts = append(textParts, text)
					}
				}
			}
			openAIMsgs = append(openAIMsgs, map[string]any{
				"role":    role,
				"content": strings.Join(textParts, "\n"),
			})
		} else {
			openAIMsgs = append(openAIMsgs, map[string]any{
				"role":    role,
				"content": fmt.Sprint(content),
			})
		}
	}
	return openAIMsgs
}

// ── Tool Conversion ──

// ConvertAnthropicToolsToOpenAI converts Anthropic tool definitions to OpenAI format.
func ConvertAnthropicToolsToOpenAI(anthropicTools []map[string]any) []map[string]any {
	var openAITools []map[string]any
	for _, tool := range anthropicTools {
		name, _ := tool["name"].(string)
		desc, _ := tool["description"].(string)
		schema, _ := tool["input_schema"].(map[string]any)

		openAITools = append(openAITools, map[string]any{
			"type": "function",
			"function": map[string]any{
				"name":        name,
				"description": desc,
				"parameters":  schema,
			},
		})
	}
	return openAITools
}

// ConvertOpenAIToolsToAnthropic converts OpenAI function definitions to Anthropic format.
func ConvertOpenAIToolsToAnthropic(openAITools []map[string]any) []map[string]any {
	var anthropicTools []map[string]any
	for _, tool := range openAITools {
		fn, _ := tool["function"].(map[string]any)
		name, _ := fn["name"].(string)
		desc, _ := fn["description"].(string)
		params, _ := fn["parameters"].(map[string]any)

		anthropicTools = append(anthropicTools, map[string]any{
			"name":         name,
			"description":  desc,
			"input_schema": params,
		})
	}
	return anthropicTools
}

// ConvertAnthropicToolUseToOpenAI converts Anthropic tool_use blocks to OpenAI function_call.
func ConvertAnthropicToolUseToOpenAI(toolUse map[string]any) map[string]any {
	name, _ := toolUse["name"].(string)
	input, _ := json.Marshal(toolUse["input"])
	return map[string]any{
		"role": "assistant",
		"function_call": map[string]any{
			"name":      name,
			"arguments": string(input),
		},
	}
}

// ConvertOpenAIFunctionCallToAnthropic converts OpenAI function_call to Anthropic tool_use.
func ConvertOpenAIFunctionCallToAnthropic(msg map[string]any) map[string]any {
	fc, _ := msg["function_call"].(map[string]any)
	name, _ := fc["name"].(string)
	var input map[string]any
	json.Unmarshal([]byte(fmt.Sprint(fc["arguments"])), &input)

	return map[string]any{
		"role":  "assistant",
		"content": []map[string]any{{
			"type":  "tool_use",
			"name":  name,
			"input": input,
		}},
	}
}

func convertOpenAIToAnthropicResponse(openAIResp map[string]any, model string) map[string]any {
	choices, _ := openAIResp["choices"].([]any)
	resp := map[string]any{
		"id":    fmt.Sprint(openAIResp["id"]),
		"model": model,
		"type":  "message",
		"role":  "assistant",
		"stop_reason": "end_turn",
	}

	var content []map[string]any
	for _, choice := range choices {
		c, _ := choice.(map[string]any)
		msg, _ := c["message"].(map[string]any)
		text, _ := msg["content"].(string)
		if text != "" {
			content = append(content, map[string]any{
				"type": "text",
				"text": text,
			})
		}
		// Convert tool calls back
		if toolCalls, ok := msg["tool_calls"].([]any); ok {
			for _, tc := range toolCalls {
				tcMap, _ := tc.(map[string]any)
				fn, _ := tcMap["function"].(map[string]any)
				name, _ := fn["name"].(string)
				var input map[string]any
				json.Unmarshal([]byte(fmt.Sprint(fn["arguments"])), &input)
				content = append(content, map[string]any{
					"type":  "tool_use",
					"name":  name,
					"input": input,
				})
			}
		}
	}

	if len(content) == 0 {
		content = []map[string]any{{"type": "text", "text": ""}}
	}
	resp["content"] = content

	// Usage
	if usage, ok := openAIResp["usage"].(map[string]any); ok {
		inputTokens, _ := toInt(usage["prompt_tokens"])
		outputTokens, _ := toInt(usage["completion_tokens"])
		resp["usage"] = map[string]any{
			"input_tokens":  inputTokens,
			"output_tokens": outputTokens,
		}
	}

	return resp
}

func convertOpenAIChunkToAnthropic(chunk map[string]any) map[string]any {
	choices, _ := chunk["choices"].([]any)
	result := map[string]any{"type": "content_block_delta"}

	if len(choices) > 0 {
		c, _ := choices[0].(map[string]any)
		delta, _ := c["delta"].(map[string]any)
		content, _ := delta["content"].(string)
		result["delta"] = map[string]any{
			"type": "text_delta",
			"text": content,
		}
	}

	return result
}

// ── Helpers ──

func toInt(v any) (int, bool) {
	switch val := v.(type) {
	case float64:
		return int(val), true
	case int:
		return val, true
	case int64:
		return int(val), true
	default:
		return 0, false
	}
}

// ── Server ──

// ServeCCSwitch creates an HTTP handler for the CC Switch proxy.
// It auto-detects Anthropic vs OpenAI format based on the request body.
func ServeCCSwitch(targetURL, apiKey, model string) http.Handler {
	cc := NewCCSwitch(targetURL, apiKey, model)

	mux := http.NewServeMux()
	mux.HandleFunc("/v1/messages", cc.HandleAnthropicRequest)
	mux.HandleFunc("/v1/chat/completions", cc.HandleOpenAIRequest)
	mux.HandleFunc("/v1/models", cc.HandleModels)

	// Catch-all for Claude Code compatibility
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if strings.Contains(r.URL.Path, "/messages") {
			cc.HandleAnthropicRequest(w, r)
			return
		}
		if strings.Contains(r.URL.Path, "/chat/completions") {
			cc.HandleOpenAIRequest(w, r)
			return
		}
		if strings.Contains(r.URL.Path, "/models") {
			cc.HandleModels(w, r)
			return
		}
		log.Printf("[CCSwitch] Unknown path: %s %s", r.Method, r.URL.Path)
		http.Error(w, "not found", 404)
	})

	return mux
}
